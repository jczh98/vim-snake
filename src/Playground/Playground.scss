.Playground {
  display: flex;
  flex-direction: row;
  background: #001d1e;
  .left-panel {
    flex: 1;
    border-right: 1px solid #0c1011;
  }
  .middle-panel {
    position: relative;
    background: #001719;
    .background {
      position: absolute;
      width: 100%;
      height: 100%;
      background-color: #001719;
      z-index: -100;
    }
    .tiles {
      background: transparent;
      z-index: 1;
      position: relative;
      overflow: hidden;
    }
  }
  .right-panel {
    flex: 1;
    border-left: 1px solid #0c1011;
  }
}
