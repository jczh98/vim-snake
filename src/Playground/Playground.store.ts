import { cast, IAnyModelType, Instance, types } from "mobx-state-tree";
import { RowStore } from "../Row";
import { snakeDirection as SnakeDirection, SnakeStore } from "../Snake";
import { ITileStore, ITileStoreSnapshotIn, TileStore } from "../Tile";

interface IPosition {
  x: number;
  y: number;
}

export interface IPlaygroundStoreConfig {
  hCount: number;
  vCount: number;
  speed: number;
  startPosition: { x: number; y: number };
}

const defaultConfig = {
  hCount: 50,
  vCount: 50,
  speed: 1,
  startPosition: {
    x: 25,
    y: 25
  }
};

export const createPlaygroundStore = (
  config: IPlaygroundStoreConfig = defaultConfig
) => {
  const { hCount, vCount, startPosition, speed } = config;

  const rows = new Array(vCount).fill(null).map((row, vi) =>
    new Array(hCount).fill(null).map(
      (tile, hi) =>
        ({
          uuid: `${hi}-${vi}`,
          position: {
            x: hi,
            y: vi
          }
        } as ITileStoreSnapshotIn)
    )
  );

  const tiles = rows.reduce((arr, row) => [...arr, ...row], []);

  const Store = types
    .model("Playground", {
      uuid: types.optional(types.identifier, "playground"),
      tiles: types.optional(types.array(TileStore), tiles),
      snake: types.optional(SnakeStore, {}),
      gameOver: types.optional(types.boolean, true),
      showVfx: true
    })
    .volatile(self => ({
      speed
    }))
    .volatile(self => {
      const state: { snakeDirection: typeof SnakeDirection.Type | null } = {
        snakeDirection: null
      };
      return state;
    })
    .views(self => ({
      get rows() {
        return rows.map((row, i) =>
          self.tiles.filter(tile => tile.position.y === i)
        );
      }
    }))
    .views(self => ({
      get hTileCount() {
        return self.rows.length;
      },
      get vTileCount() {
        const row = self.rows.length === 0 ? null : self.rows[0];
        return row ? row.length : 0;
      },
      get speedInterval() {
        return Math.floor(1000 / self.speed);
      },
      get score() {
        return self.snake.tiles.length;
      },
      get allTiles() {
        return self.rows.reduce(
          (arr, row) => [...arr, ...row],
          [] as ITileStore[]
        );
      }
    }))
    .actions(self => ({
      setGameOver() {
        self.gameOver = true;
        self.snakeDirection = null;
      },
      setSnake(position: IPosition) {
        const { x, y } = position;
        self.snake.eatTile(self.rows[y][x]);
      },
      toggleShowVfx() {
        self.showVfx = !self.showVfx;
      },
      setSnakeDirection(direction: typeof SnakeDirection.Type) {
        const { snakeDirection } = self;
        switch (direction) {
          case "Left":
            if (snakeDirection === "Right") return;
            break;
          case "Right":
            if (snakeDirection === "Left") return;
            break;
          case "Down":
            if (snakeDirection === "Up") return;
            break;
          case "Up":
            if (snakeDirection === "Down") return;
            break;
          default:
        }
        self.snakeDirection = direction;
      },
      resetSnake() {
        self.snake.reset();
        this.setSnake(startPosition);
      },
      restartGame() {
        self.allTiles.forEach(tile => tile.setState("Empty"));
        self.snakeDirection = null;
        this.resetSnake();
        this.addFood();
        self.gameOver = false;
      },
      addFood(
        position: IPosition = createRandomPosition(
          self.hTileCount,
          self.vTileCount
        )
      ) {
        const { x, y } = position;
        const tile = self.rows[y][x];
        if (tile.state !== "Empty") {
          this.addFood();
        }
        tile.setState("Food");
      },
      moveSnake() {
        if (self.gameOver) return;
        switch (self.snakeDirection) {
          case "Right":
            self.snake.moveRight();
            break;
          case "Left":
            self.snake.moveLeft();
            break;
          case "Down":
            self.snake.moveDown();
            break;
          case "Up":
            self.snake.moveUp();
            break;
          default:
            break;
        }
      }
    }));
  return Store;
};

function createRandomPosition(hTileCount: number, vTileCount: number) {
  return {
    x: Math.floor(Math.random() * Math.floor(hTileCount)),
    y: Math.floor(Math.random() * Math.floor(vTileCount))
  };
}

const PlaygroundStore = createPlaygroundStore();

export type PlaygroundStoreType = typeof PlaygroundStore;

export interface IPlaygroundStore extends Instance<PlaygroundStoreType> {}
