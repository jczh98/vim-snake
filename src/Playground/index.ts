export * from "./Playground";
export * from "./Playground.store";
