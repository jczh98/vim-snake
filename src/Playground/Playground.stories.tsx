import { number, withKnobs } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";
import React from "react";
import { Playground } from "./Playground";
import { createPlaygroundStore } from "./Playground.store";

const PlaygroundStore = createPlaygroundStore();

const stories = storiesOf("Playground", module);

stories.addDecorator(withKnobs);

stories.add("basic", () => {
  return <Playground store={PlaygroundStore.create({})} />;
});

stories.add("with snake", () => {
  const store = PlaygroundStore.create({});
  return <Playground store={store} />;
});
