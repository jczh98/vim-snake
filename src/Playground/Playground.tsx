import { action, computed } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import { InfoPanel } from "../InfoPanel";
import { Row } from "../Row/Row";
import { SettingsPanel } from "../SettingsPanel";
import "./Playground.scss";
import { IPlaygroundStore } from "./Playground.store";
interface IProps {
  store: IPlaygroundStore;
}

@observer
export class Playground extends React.Component<IProps> {
  @computed
  private get tiles() {
    const { store } = this.props;
    const { rows } = this.props.store;
    return rows.map((row, i) => (
      <Row key={i} tiles={row} playgroundStore={store} />
    ));
  }
  private moveTimeout!: number;
  public componentDidMount() {
    window.addEventListener("keydown", this.onKeyDown);
    this.moveTimeout = window.setInterval(
      this.props.store.moveSnake,
      this.props.store.speedInterval
    );
  }
  public componentWillMount() {
    window.removeEventListener("keydown", this.onKeyDown);
    window.clearInterval(this.moveTimeout);
  }
  public render() {
    const { store } = this.props;
    return (
      <div className="Playground">
        <div className="left-panel">
          <SettingsPanel store={store} />
        </div>
        <div className="middle-panel">
          <div className="background" />
          <div className="tiles">{this.tiles}</div>
        </div>
        <div className="right-panel">
          <InfoPanel store={store} />
        </div>
      </div>
    );
  }

  @action.bound
  private onKeyDown({ key }: KeyboardEvent) {
    switch (key) {
      case "ArrowLeft":
        this.props.store.setSnakeDirection("Left");
        break;
      case "ArrowRight":
        this.props.store.setSnakeDirection("Right");
        break;
      case "ArrowUp":
        this.props.store.setSnakeDirection("Up");
        break;
      case "ArrowDown":
        this.props.store.setSnakeDirection("Down");
        break;
      default:
        break;
    }
  }
}
