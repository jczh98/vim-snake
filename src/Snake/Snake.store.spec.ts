import { createPlaygroundStore, IPlaygroundStoreConfig } from "../Playground";

describe("single tile snake", () => {
  function createPlayground(config: Partial<IPlaygroundStoreConfig> = {}) {
    const defaultConfig = {
      hCount: 10,
      vCount: 10,
      speed: 1,
      startPosition: { x: 0, y: 0 }
    };
    const PlaygroundStore = createPlaygroundStore({
      ...defaultConfig,
      ...config
    });
    const playground = PlaygroundStore.create({});
    playground.restartGame();
    const { snake } = playground;
    return { playground, snake };
  }

  it("create Snake", () => {
    const { playground, snake } = createPlayground();
    const tile = playground.rows[0][0];
    expect(snake.tiles).toEqual([tile]);
  });

  it("move right", () => {
    const { playground, snake } = createPlayground();
    playground.setSnakeDirection("Right");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[0][1]]);
  });

  it("reset snake to begining of row on move right on the edge", () => {
    const { playground, snake } = createPlayground({
      hCount: 10,
      startPosition: { x: 9, y: 0 }
    });
    playground.setSnakeDirection("Right");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[0][0]]);
  });

  it("move left", () => {
    const { playground, snake } = createPlayground({
      hCount: 10,
      startPosition: { x: 9, y: 0 }
    });
    playground.setSnakeDirection("Left");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[0][8]]);
  });

  it("reset snake to end of row on move left on the edge", () => {
    const { playground, snake } = createPlayground({
      hCount: 10,
      startPosition: { x: 0, y: 0 }
    });
    playground.setSnakeDirection("Left");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[0][9]]);
  });

  it("move down", () => {
    const { playground, snake } = createPlayground({
      vCount: 10,
      startPosition: { x: 0, y: 0 }
    });
    playground.setSnakeDirection("Down");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[1][0]]);
  });

  it("reset snake to top of column on move down on the edge", () => {
    const { playground, snake } = createPlayground({
      vCount: 10,
      startPosition: { x: 0, y: 9 }
    });
    playground.setSnakeDirection("Down");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[0][0]]);
  });

  it("move up", () => {
    const { playground, snake } = createPlayground({
      vCount: 10,
      startPosition: { x: 0, y: 9 }
    });
    playground.setSnakeDirection("Up");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[8][0]]);
  });

  it("reset snake to bottom of column on move up on the edge", () => {
    const { playground, snake } = createPlayground({
      vCount: 10,
      startPosition: { x: 0, y: 0 }
    });
    playground.setSnakeDirection("Up");
    playground.moveSnake();
    expect(snake.tiles).toEqual([playground.rows[9][0]]);
  });
});

describe("three tile size snake", () => {
  function createPlayground(config: Partial<IPlaygroundStoreConfig> = {}) {
    const defaultConfig = {
      hCount: 10,
      vCount: 10,
      speed: 1,
      startPosition: { x: 0, y: 0 }
    };
    const PlaygroundStore = createPlaygroundStore({
      ...defaultConfig,
      ...config
    });
    const playground = PlaygroundStore.create({});
    playground.resetSnake();
    const { snake } = playground;
    snake.eatTile(playground.rows[0][1]);
    snake.eatTile(playground.rows[0][2]);
    return { playground, snake };
  }
  it("set all tiles", () => {
    const { snake } = createPlayground();

    expect(snake.tiles.length).toBe(3);
    expect(snake.tiles).toMatchSnapshot();
  });
  it("move right", () => {
    const { snake } = createPlayground();
    expect(snake.tiles).toMatchSnapshot();
    snake.moveRight();
    expect(snake.tiles).toMatchSnapshot();
  });

  it("eat tile", () => {
    const { snake, playground } = createPlayground();
    playground.addFood({ x: 3, y: 0 });
    expect(snake.tiles.length).toBe(3);
    snake.moveRight();
    expect(snake.tiles.length).toBe(4);
    expect(snake.tiles).toMatchSnapshot();
  });

  it("change direction will follow path", () => {
    const { snake, playground } = createPlayground();
    expect(snake.tiles).toMatchSnapshot();
    snake.moveRight();
    expect(snake.tiles).toMatchSnapshot();
    snake.moveDown();
    expect(snake.tiles[0].position).toEqual({ x: 3, y: 1 });
    expect(snake.tiles).toMatchSnapshot();
  });
});
