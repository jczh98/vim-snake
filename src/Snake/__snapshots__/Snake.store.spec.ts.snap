// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`three tile size snake change direction will follow path 1`] = `
Array [
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "2-0",
  },
  Object {
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "1-0",
  },
  Object {
    "position": Object {
      "x": 0,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "0-0",
  },
]
`;

exports[`three tile size snake change direction will follow path 2`] = `
Array [
  Object {
    "position": Object {
      "x": 3,
      "y": 0,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "3-0",
  },
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "2-0",
  },
  Object {
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "1-0",
  },
]
`;

exports[`three tile size snake change direction will follow path 3`] = `
Array [
  Object {
    "position": Object {
      "x": 3,
      "y": 1,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "3-1",
  },
  Object {
    "position": Object {
      "x": 3,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "3-0",
  },
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "2-0",
  },
]
`;

exports[`three tile size snake eat tile 1`] = `
Array [
  Object {
    "position": Object {
      "x": 3,
      "y": 0,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "3-0",
  },
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "2-0",
  },
  Object {
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "1-0",
  },
  Object {
    "position": Object {
      "x": 0,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "0-0",
  },
]
`;

exports[`three tile size snake move right 1`] = `
Array [
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "2-0",
  },
  Object {
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "1-0",
  },
  Object {
    "position": Object {
      "x": 0,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "0-0",
  },
]
`;

exports[`three tile size snake move right 2`] = `
Array [
  Object {
    "position": Object {
      "x": 3,
      "y": 0,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "3-0",
  },
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "2-0",
  },
  Object {
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "1-0",
  },
]
`;

exports[`three tile size snake set all tiles 1`] = `
Array [
  Object {
    "position": Object {
      "x": 2,
      "y": 0,
    },
    "snakeHead": true,
    "state": "Snake",
    "uuid": "2-0",
  },
  Object {
    "position": Object {
      "x": 1,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "1-0",
  },
  Object {
    "position": Object {
      "x": 0,
      "y": 0,
    },
    "snakeHead": false,
    "state": "Snake",
    "uuid": "0-0",
  },
]
`;
